Write path fixes, integration tests

dead10ck opened the pull request:
Well, I did not expect this PR to get this large, I'm so sorry about that. This PR started as an attempt to fix #1575, and along the way, I discovered that there was a larger problem: Document writes are submitted to a single job pool that all run unordered. I suspected that this could lead to file write inconsistencies if the user submitted multiple writes on the same document at the same time. I took this opportunity to continue work on the integration testing harness that @archseer started a while back. I added an integration test (see below) that confirms that indeed the file writes happen not only out of order, but simtultaneously, leading to the file being clobbered. In a test where we issue a thousand writes that each change the entire file to consist only of the iteration number (i.e., 1, 2, 3, ..., 1000), we can see that the file ends up with a seemingly random jumble:

```
failures:

---- integration::write::test_write_concurrent stdout ----
thread 'integration::write::test_write_concurrent' panicked at 'assertion failed: `(left == right)`
  left: `"1000"`,
 right: `"4010"`', helix-term/tests/integration/write.rs:66:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace


failures:
    integration::write::test_write_concurrent
```

This test is now passing with the changes to the write path.

Below is an explanation of some of my decisions in this PR. Opinions and suggestions are welcome!

## Document writes

The approach taken was:

1. Rather than submitting a future to the general job pool, each Document manages its own queue of writes with an mpsc channel. When a user submits a write, this merely queues up a write request on the internal channel.
1. Document writes emit a new special event type that communicates the revision number that was written and the path. The main event loop is responsible for polling the futures, which prompt the runtime to carry out the actual writes. When writes finish, the main event loop will update the document's last saved revision and the path if the write was successful.
1. All events that originate from the Editor were consolidated into one enum type. This was discussed in the chat room, and it was originally expressed that we could do this incrementally over time, but the borrow checker intervened with this plan: since the event polls require a mutable borrow, and we can only do this once within the event loop, we have to do this all at once.

Incidentally, this allowed fixes for two unrelated bugs that presently exist:

* When a document write fails, the modified flag is still reset. By sending the revision number in the write event, the current revision number and the last saved revision number need not be the same at any given point in time. This would, e.g., allow large writes to carry out in the background while the user makes other edits at the same time, and correctly keep the modified flag even after the background write succeeds.
* In a write command that sets a new path, if the write fails, the path is still changed to the  argument path. Now, it will not update the path unless the write succeeds.
  
Also, the first attempt at fixing #1575 was done by adding fallibility to the job pool. Presently, if you join on all the jobs, any errors are discarded.

Since I was already here, I attempted to fix #1633 as well, and it kind of works, except that the status message quickly gets overwritten by LSP progress messages. I'm not quite sure how to fix this, but I think that can happen in another PR, as this one is quite big enough already.

  
## Integration tests

I was already working on an integration testing branch in the background for a little while, and I thought since this is such a critical code path, it warrants introducing the integration testing harness at the same time. This is very much a "just getting the ball rolling" approach, and anything is up for discussion. I fully expect the integration test framework to change over time.

Some things to note about this implementation:

* In order to facilitate easier testing, I moved the toml config parsing back into the main function. This undoes the move that was done in #1249.
* The integration tests require shutting off actual rendering, so to do this, a feature branch was introduced to disable rendering for the integration tests. Tests are run by doing `cargo test --features integration`
* We need some way to run the Application with some starting document and input keys, and then inspect the internal state while the Application is still running. We also need to run the actual main event loop in order to handle more things than just the key press event—we need to drive document writes, config changes, and eventually LSP messages, if and when any tests are added for these things. The problem is the main event loop doesn't exit until all documents are closed, and then it discards a lot of internal state.

  So in order to inspect internal state, I first just did this with a simple timeout, but it quickly became clear that this is problematic, as the amount of time that a test finishes can differ dramatically, and it can lead to non-determinism: if the host system becomes busy for some reason, tests fail because they time out too soon, even when there isn't a problem. Additionally, we can't just use very long timeouts as this would artificially make it take a very long time for each integration test to finish, even if the actual work they did was done very quickly.

  So I came up with the idea that we could piggy-back on the idle timer: since the event loop is using a biased event loop, and we submit all key strokes ahead of actually running the Application, the idle timer should only fire when every key has been handled. To implement this, the event loop was split into one which exits after a successful idle timer handle, and the main call site simply loops over this as well. Additionally, the idle timer is reset every time any event gets handled successfully. This lets the application signal when it is done processing the test key strokes, and leaves the internal state open for inspection.

Another bug was discovered along the way while writing these integration tests: when a document is first opened, the cursor has the initial state of (0, 0), and is not updated to reflect the document's contents until the first cursor movement. This led to some weird bugs where the selection did not end up in the right direction after an edit. As far as I have been able to find, this does not cause any bugs in the actual text transformation, so it's likely this did not cause any real harm to users, but nonetheless, it was making testing very confusing until it was fixed. Now, if there is text in the opened document, the cursor is placed over the whole first grapheme cluster.

Whew! I apologize again for the size of this PR. I hope we can get a few eyes on this, as it is in the critical write path. Ideally, we could get people willing to test it before merging as well, in order to catch any possible bugs.

Open questions:

* How do we handle the feature flag? Should we simply leave it as is, and require adding the flag to run integration tests, or should we alias `test` to do `test --features integration`? The latter would help by not requiring contributors to remember that you have to do this to run the integration tests, but has the issue that eventually, we may have tests that depend upon external state, such as an LSP on the build system. Perhaps there could be an xtask or something? I'm open to suggestions.
* When we do have tests that need consistent external state, how do we do this? A docker container?

dead10ck commented:
Not really sure what's going on with the EAGAIN errors. Trying to dig into it. I'm guessing it has something to do with the terminal in the build environment.

the-mikedavis commented:
Would it be possible to split the PR up into two - one for the integration tests and one to fix document writes (probably based on the integration testing one)? It is quite large as you say 😅

dead10ck commented:
Sure, I have some of the changes mixed up, but I'll see if I can rebase and spilt them up

pickfire commented on helix-term/tests/integration.rs:
```diff
@@ -0,0 +1,31 @@
+
+    use self::helpers::*;
+
+    #[tokio::test]
+    async fn hello_world() -> anyhow::Result<()> {
+        test_key_sequence_text_result(
+            Args::default(),
+            Config::default(),
+            ("#[\n|]#", "ihello world<esc>", "hello world#[|\n]#"),
+        )
+        .await?;
+
+        Ok(())
+    }
```
Interesting, now we have integration test

dead10ck replied:
Hm, I kind of liked the explicitness of seeing that this is a test case with the default config. But yeah, it's seeming like most of the test cases are with the default, so maybe another helper would be good. I kind of feel like there are more than I would like already though. Maybe I can make a variadic macro or something.

pickfire commented:
I think you accidentally included `out` log file.

pickfire commented on helix-core/src/selection.rs:
```diff
@@ -339,6 +339,14 @@ impl Range {
     }
+
+    /// Returns true if this Range covers a single grapheme in the given text
+    pub fn is_single_grapheme(&self, doc: RopeSlice) -> bool {
+        let mut graphemes = RopeGraphemes::new(doc.slice(self.from()..self.to()));
```
We usually call it `text` instead of `doc`.

pickfire commented on helix-term/src/application.rs:
```diff
@@ -53,32 +62,39 @@ pub struct Application {
+
+#[cfg(feature = "integration")]
+fn setup_integration_logging() {
+    let level = std::env::var("HELIX_LOG_LEVEL")
+        .map(|lvl| lvl.parse().unwrap())
+        .unwrap_or(log::LevelFilter::Info);
```
It would be nice if we can have log env kind like of filter rather than showing everything. Like `RUST_LOG=helix-term:info`.

dead10ck replied:
Yeah I thought about that too, but that would mean introducing a second logger.

pickfire commented on helix-term/src/application.rs:
```diff
@@ -265,11 +290,56 @@ impl Application {
-                    // idle timeout
-                    self.editor.clear_idle_timer();
-                    self.handle_idle_timeout();
+                event = self.editor.wait_event() => {
```
Does this need bias? Otherwise the event might fill up the other events.

dead10ck replied:
I think it's best to keep it biased. I think it's best to prioritize key presses so the user has the best experience, otherwise e.g. if there are a bunch of LSP messages coming through, it could introduce a perceived lag in the responsiveness. Likewise, we don't want to delay file writes because of debugger events or config changes.

pickfire commented on helix-term/src/application.rs:
```diff
@@ -365,23 +435,71 @@ impl Application {
+
+        let doc = doc.unwrap();
+
+        debug!(
+            "document {:?} saved with revision {}",
+            doc.path(),
+            doc_save_event.revision
+        );
```
Hmm, I wonder why editors don't show user the revision, if it did then users are able to use time travel easily. Like `:ear 1f`.

archseer commented:
Can you rebase this now? :)

archseer commented on helix-term/src/application.rs:
```diff
@@ -305,20 +297,49 @@ impl Application {
+                        }
+                        EditorEvent::ConfigEvent(event) => {
+                            self.handle_config_events(event);
+                            self.render();
+                        }
+                        EditorEvent::LanguageServerMessage((id, call)) => {
+                            self.handle_language_server_message(call, id).await;
+                            // limit render calls for fast language server messages
+                            let last = self.editor.language_servers.incoming.is_empty();
+
+                            if last || self.last_render.elapsed() > LSP_DEADLINE {
+                                self.render();
+                                self.last_render = Instant::now();
+                            }
+                        }
+                        EditorEvent::DebuggerEvent(payload) => {
+                            let needs_render = self.editor.handle_debugger_message(payload).await;
+                            if needs_render {
+                                self.render();
+                            }
+                        }
+                        EditorEvent::IdleTimer => {
+                            self.editor.clear_idle_timer();
+                            self.handle_idle_timeout();
+
+                            #[cfg(feature = "integration")]
+                            {
+                                idle_handled = true;
+                            }
+                        }
                     }
```
To reduce `select!` complexity, let's move this into a submethod and mark it as `inline(always)`

pickfire commented on helix-term/src/application.rs:
```diff
@@ -822,19 +892,56 @@ impl Application {
         };
 
-        Ok(())
+        for doc in self.editor.documents_mut() {
+            if let Some(save_result) = doc.close().await {
+                result = match save_result {
+                    Ok(_) => result,
+                    Err(err) => {
+                        if let Some(path) = doc.path() {
+                            log::error!(
+                                "Error saving document '{}': {}",
+                                path.to_string_lossy(),
+                                err
+                            );
+                        }
+                        Err(err)
+                    }
+                };
+            }
+        }
```
Is this awaiting for each item one by one or it is runned asynchronously all of them? I haven't written async/await in some time and I am not sure the behavior here.

dead10ck replied:
I believe this will be one by one. This code is only called during the shutdown process though, so this would only matter if the user did something pathological like open a few dozen multigigabyte files, changed them all, and then did a `write-quit-all`.

pickfire commented on helix-view/src/editor.rs:
```diff
@@ -862,6 +876,19 @@ impl Editor {
         Ok(id)
     }
 
+    /// Closes every document and view, which will ultimately shut down the app.
+    // pub async fn close(&mut self) -> anyhow::Result<()> {
+    //     let docs: Vec<_> = self.documents().map(|doc| doc.id()).collect();
+
+    //     // closing all documents also closes all views, which will shut
+    //     // down the app
+    //     for doc in docs {
+    //         self.close_document(doc, false).await?;
+    //     }
+
+    //     Ok(())
+    // }
```
Do we need to keep this?

dead10ck replied:
Oops, don't think so, will remove.

dead10ck commented:
The issue of the auto format changing the revision number is proving to be kind of a tough nut to crack. I tried making it work by having the auto format callback kick off the save, but the problem here is that the callback actually needs to run, and the current set up is such that the main event loop is the only place callbacks are handled, so write-quit doesn't end up actually doing a write. I tried adding logic to handle callbacks during `Jobs::finish`, but we don't have the compositor available to us at this point, since that is a terminal specific thing.

I'm also considering if we could just somehow change the jobs so that any particular job can track when its callback has been handled, and then the write would just wait for that.

Progress has been slowed down recently due to personal stuff, but I am still trying to work through it when I can to get this PR done.

dead10ck commented:
Ok, I managed to get something working that waits for auto format and save to finish on `write-quit`

It turns the callback type into an enum to allow for callbacks that don't have a compositor and/or editor, and then basically allows you to tell `Jobs::finish` to ignore callbacks for which you don't give it a necessary reference (i.e. if you don't pass in a Compositor, ignore any callbacks that requires a compositor).

It's a bit of a hack, but it's simpler than some of the other ideas I was able to think of.

I'm open to other suggestions as well.

the-mikedavis commented:
I found a panic that happens when you quit immediately following a save when you have an active language server that auto-formats on save:

1. `cargo run helix-term/src/ui/editor.rs`
2. `:w`
3. `:q`

I think the `:q` needs to be entered before the auto-formatting response is received.

<details><summary>backtrace...</summary>

```
thread 'main' panicked at 'called `Option::unwrap()` on a `None` value', helix-view/src/tree.rs:274:29
stack backtrace:
   0:     0x56463b08e67d - std::backtrace_rs::backtrace::libunwind::trace::h22893a5306c091b4
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/../../backtrace/src/backtrace/libunwind.rs:93:5
   1:     0x56463b08e67d - std::backtrace_rs::backtrace::trace_unsynchronized::h29c3bc6f9e91819d
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x56463b08e67d - std::sys_common::backtrace::_print_fmt::he497d8a0ec903793
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/sys_common/backtrace.rs:66:5
   3:     0x56463b08e67d - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h9c2a9d2774d81873
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/sys_common/backtrace.rs:45:22
   4:     0x56463b0b5f4c - core::fmt::write::hba4337c43d992f49
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/core/src/fmt/mod.rs:1194:17
   5:     0x56463b0878b1 - std::io::Write::write_fmt::heb73de6e02cfabed
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/io/mod.rs:1655:15
   6:     0x56463b0905f5 - std::sys_common::backtrace::_print::h63c8b24acdd8e8ce
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/sys_common/backtrace.rs:48:5
   7:     0x56463b0905f5 - std::sys_common::backtrace::print::h426700d6240cdcc2
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/sys_common/backtrace.rs:35:9
   8:     0x56463b0905f5 - std::panicking::default_hook::{{closure}}::hc9a76eed0b18f82b
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/panicking.rs:295:22
   9:     0x56463b0902a9 - std::panicking::default_hook::h2e88d02087fae196
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/panicking.rs:314:9
  10:     0x564639a491e3 - <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call::hb10f8f69f0ddaaee
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/alloc/src/boxed.rs:1875:9
  11:     0x564639a19808 - helix_term::application::Application::run::{{closure}}::{{closure}}::h3801ef15e785355d
                               at /home/michael/src/helix/hx/helix-term/src/application.rs:901:13
  12:     0x56463b090d00 - std::panicking::rust_panic_with_hook::habfdcc2e90f9fd4c
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/panicking.rs:702:17
  13:     0x56463b090af9 - std::panicking::begin_panic_handler::{{closure}}::he054b2a83a51d2cd
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/panicking.rs:586:13
  14:     0x56463b08eb34 - std::sys_common::backtrace::__rust_end_short_backtrace::ha48b94ab49b30915
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/sys_common/backtrace.rs:138:18
  15:     0x56463b090869 - rust_begin_unwind
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/panicking.rs:584:5
  16:     0x56463b0b42c3 - core::panicking::panic_fmt::h366d3a309ae17c94
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/core/src/panicking.rs:143:14
  17:     0x56463b0b410d - core::panicking::panic::h8705e81f284be8a5
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/core/src/panicking.rs:48:5
  18:     0x56463a3c6dfc - core::option::Option<T>::unwrap::hfb651dc565a95d85
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/core/src/option.rs:755:21
  19:     0x56463a3db6be - helix_view::tree::Tree::get::h9be0133c774319e5
                               at /home/michael/src/helix/hx/helix-view/src/tree.rs:274:9
  20:     0x564639b678b0 - helix_term::commands::make_format_callback::{{closure}}::{{closure}}::h35c41125ed72bca4
                               at /home/michael/src/helix/hx/helix-term/src/commands.rs:2479:23
  21:     0x564639c52998 - core::ops::function::FnOnce::call_once{{vtable.shim}}::haa758735588049f4
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/core/src/ops/function.rs:227:5
  22:     0x564639c45bfa - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::hb9b36a8e823b7b97
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/alloc/src/boxed.rs:1861:9
  23:     0x564639df971d - helix_term::job::Jobs::finish::{{closure}}::h94411ea9371700c4
                               at /home/michael/src/helix/hx/helix-term/src/job.rs:128:33
  24:     0x564639f6b4fe - <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll::hb734ba7cf1d0f750
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/core/src/future/mod.rs:91:19
  25:     0x564639a19baf - helix_term::application::Application::close::{{closure}}::h50a58e8c7c952179
                               at /home/michael/src/helix/hx/helix-term/src/application.rs:919:72
  26:     0x564639a78c99 - <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll::h941c8e4690160904
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/core/src/future/mod.rs:91:19
  27:     0x564639a194c9 - helix_term::application::Application::run::{{closure}}::h685593b0af45da56
                               at /home/michael/src/helix/hx/helix-term/src/application.rs:906:21
  28:     0x564639a7946b - <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll::hec3dd69f9ce717f3
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/core/src/future/mod.rs:91:19
  29:     0x564639a3273d - hx::main_impl::{{closure}}::h954da21fc956f2d1
                               at /home/michael/src/helix/hx/helix-term/src/main.rs:142:53
  30:     0x564639a7911b - <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll::hccd7dc1c48f19f56
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/core/src/future/mod.rs:91:19
  31:     0x564639a2c8cc - tokio::park::thread::CachedParkThread::block_on::{{closure}}::h9fce7fe91ecde133
                               at /home/michael/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.20.1/src/park/thread.rs:263:54
  32:     0x564639a35dfa - tokio::coop::with_budget::{{closure}}::h5496a9b2ab317313
                               at /home/michael/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.20.1/src/coop.rs:102:9
  33:     0x564639a2d168 - std::thread::local::LocalKey<T>::try_with::h3beef407bce4b31e
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/thread/local.rs:442:16
  34:     0x564639a2ca4c - std::thread::local::LocalKey<T>::with::h5321acaa83343402
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/thread/local.rs:418:9
  35:     0x564639a2c242 - tokio::coop::with_budget::h06e05fd16a115c8a
                               at /home/michael/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.20.1/src/coop.rs:95:5
  36:     0x564639a2c242 - tokio::coop::budget::had8f4b1b93b3f2c1
                               at /home/michael/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.20.1/src/coop.rs:72:5
  37:     0x564639a2c242 - tokio::park::thread::CachedParkThread::block_on::h87d406f798c9e3a8
                               at /home/michael/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.20.1/src/park/thread.rs:263:31
  38:     0x564639a37d55 - tokio::runtime::enter::Enter::block_on::h6006bfc704f6ecd1
                               at /home/michael/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.20.1/src/runtime/enter.rs:152:13
  39:     0x564639a6b068 - tokio::runtime::thread_pool::ThreadPool::block_on::he30ecd538d338b64
                               at /home/michael/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.20.1/src/runtime/thread_pool/mod.rs:90:9
  40:     0x564639a7b306 - tokio::runtime::Runtime::block_on::h5800ac6a89a0990b
                               at /home/michael/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.20.1/src/runtime/mod.rs:484:43
  41:     0x564639a2b62d - hx::main_impl::h1b84079ec26e0a46
                               at /home/michael/src/helix/hx/helix-term/src/main.rs:144:5
  42:     0x564639a2b47e - hx::main::h058367f4d46396e7
                               at /home/michael/src/helix/hx/helix-term/src/main.rs:37:21
  43:     0x564639a59e0b - core::ops::function::FnOnce::call_once::h8031714db130c54d
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/core/src/ops/function.rs:227:5
  44:     0x564639a2e2ae - std::sys_common::backtrace::__rust_begin_short_backtrace::hacdace9adcecea47
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/sys_common/backtrace.rs:122:18
  45:     0x564639a45761 - std::rt::lang_start::{{closure}}::ha07536b68a461eb7
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/rt.rs:145:18
  46:     0x56463b08157e - core::ops::function::impls::<impl core::ops::function::FnOnce<A> for &F>::call_once::had4f69b3aefb47a8
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/core/src/ops/function.rs:259:13
  47:     0x56463b08157e - std::panicking::try::do_call::hf2ad5355fcafe775
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/panicking.rs:492:40
  48:     0x56463b08157e - std::panicking::try::h0a63ac363423e61e
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/panicking.rs:456:19
  49:     0x56463b08157e - std::panic::catch_unwind::h18088edcecb8693a
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/panic.rs:137:14
  50:     0x56463b08157e - std::rt::lang_start_internal::{{closure}}::ha7dad166dc711761
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/rt.rs:128:48
  51:     0x56463b08157e - std::panicking::try::do_call::hda0c61bf3a57d6e6
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/panicking.rs:492:40
  52:     0x56463b08157e - std::panicking::try::hbc940e68560040a9
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/panicking.rs:456:19
  53:     0x56463b08157e - std::panic::catch_unwind::haed0df2aeb3fa368
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/panic.rs:137:14
  54:     0x56463b08157e - std::rt::lang_start_internal::h9c06694362b5b80c
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/rt.rs:128:20
  55:     0x564639a45730 - std::rt::lang_start::h44af8e86ec90ecff
                               at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/rt.rs:144:17
  56:     0x564639a2b6dc - main
  57:     0x7f4ec38d3237 - __libc_start_call_main
  58:     0x7f4ec38d32f5 - __libc_start_main_impl
  59:     0x564639a090f1 - _start
                               at /build/glibc-2.34/csu/../sysdeps/x86_64/start.S:116
  60:                0x0 - <unknown>
```

</details>

dead10ck commented:
Nice, good catch @the-mikedavis! This makes sense. It looks like the format callback needs the view to apply the LSP edits, but the [shutdown (1)](https://github.com/helix-editor/helix/blob/cb54ed3331b4449b5372a30255ef59c06d735283/helix-term/src/application.rs#L272) [sequence(2)](https://github.com/helix-editor/helix/blob/cb54ed3331b4449b5372a30255ef59c06d735283/helix-view/src/editor.rs#L1109) is [initiated by closing the view](https://github.com/helix-editor/helix/blob/cb54ed3331b4449b5372a30255ef59c06d735283/helix-term/src/commands/typed.rs#L34), so by the time we execute the job, the view is gone.

This was fixed in write-quit by blocking on the jobs finishing before quitting, but it looks like this will have to happen even for a regular quit.

I suspect the only reason this isn't common already is because this job wasn't marked as wait before quit before, but now it is, so this code never got a chance to execute even though it would panic the same way.

dead10ck commented:
Ok, I think I fixed it

David-Else commented:
Awesome! Does it fix: https://github.com/helix-editor/helix/issues/1639 ? Hoping to use ZZ to quit and save when the bug is gone: https://github.com/helix-editor/helix/issues/2041

pickfire commented on helix-view/src/editor.rs:
```diff
@@ -1147,7 +1172,7 @@ impl Editor {
     ) -> Result<(), tokio::time::error::Elapsed> {
         tokio::time::timeout(
-            Duration::from_millis(timeout.unwrap_or(500)),
+            Duration::from_millis(timeout.unwrap_or(10000)),
```
The comment above is outdated when you changed this to 10000.

pickfire commented on helix-term/src/application.rs:
```diff
@@ -827,19 +901,56 @@ impl Application {
+            }
+        }
+
+        match self.editor.close_language_servers(None).await {
```
I wonder if we still need the argument given that this is not used in other places, might as well just drop it.

dead10ck replied:
Hmm, being able to configure an LSP shutdown timeout seems useful though, especially given we've had recent bug reports like #1639. I think we just need to do the work of adding a config option for some of the timeouts for LSP requests.

pickfire commented on helix-term/src/compositor.rs:
```diff
@@ -27,6 +27,24 @@ pub struct Context<'a> {
+            helix_lsp::block_on(self.jobs.finish(Some(self.editor), None))
+        })?;
+
+        for doc in &mut self.editor.documents.values_mut() {
+            tokio::task::block_in_place(|| helix_lsp::block_on(doc.try_flush_saves()))
+                .map(|result| result.map(|_| ()))
+                .unwrap_or(Ok(()))?;
+        }
```
Should we try to save all the files rather than bailing out on the first error? But I think this should be fine for now since it's rare to save multiple files and to fail.

dead10ck replied:
Possibly, although I don't know if it's worth the added complexity of the error handling at the call site. You'd have to do something like return a `Vec<Result<_>>`.

pickfire replied:
I guess it is fine for now then.

pickfire reviewed (approved):
Looks good to me.

dead10ck commented:
Thanks for taking a look @pickfire ! Would you be willing to dogfood this build for a bit of time to see if you notice any issues in practice? If we have the ability, it would be prudent to field test a change of this significance before merging to master for everyone.

pickfire commented:
@dead10ck Sure, will try this out when doing normal editing rather than master branch.

archseer commented on helix-term/src/commands/typed.rs:
```diff
@@ -71,8 +76,10 @@ fn buffer_close_by_ids_impl(
+    log::debug!("closing buffers: {:?}", doc_ids);
+
+    for doc_id in doc_ids {
+        tokio::task::block_in_place(|| helix_lsp::block_on(editor.close_document(*doc_id, force)))?;
```
Why the double `block_on` here? We should avoid blocking if possible since I've managed to get deadlocks with `block_on` + `I/O` before

dead10ck replied:
See the [docs](https://docs.rs/tokio/1.20.1/tokio/task/fn.block_in_place.html)

>Runs the provided blocking function on the current thread without blocking the executor.
>
>In general, issuing a blocking call or performing a lot of compute in a future without yielding is problematic, as it may prevent the executor from driving other tasks forward. Calling this function informs the executor that the currently executing task is about to block the thread, so the executor is able to hand off any other tasks it has to a new worker thread before that happens. See the CPU-bound tasks and blocking code section for more information.

So the outer `block_in_place` does not block, but just runs the provided blocking function and tells the executor to move other tasks off this thread, specifically to avoid deadlocking, like you said. I actually ran into this myself, which is how I found this helper.

Avoiding blocking will be difficult to say the least. By the end of the command, the buffer needs to be closed, so we have to make sure all the writes on the document are done. This is similar to shutting down the application: you have to make sure all the writes are done before quitting, like we do now by waiting on the jobs before quitting. You *must* block at some point or another.

archseer commented on helix-term/src/compositor.rs:
```diff
@@ -27,6 +27,24 @@ pub struct Context<'a> {
+    /// operations for all documents.
+    pub fn block_try_flush_writes(&mut self) -> anyhow::Result<()> {
+        tokio::task::block_in_place(|| {
+            helix_lsp::block_on(self.jobs.finish(Some(self.editor), None))
```
This has a timeout, right?

dead10ck replied:
Currently, this is only controlled by each future/job. For example, the LSP client always wraps the requests in a timeout. But there is no globally enforced timeout. This is the case as of today as well.

archseer commented on helix-term/src/job.rs:
```diff
@@ -93,13 +100,41 @@ impl Jobs {
+                        #[allow(clippy::needless_option_as_deref)]
+                        match callback {
+                            Callback::EditorCompositor(call)
+                                if editor.is_some() && compositor.is_some() =>
```
Why not just always callback with both? This seems error prone since it's possible to call `finish()` with `editor` as `None` and then the callback will just silently be skipped

archseer commented on helix-view/src/editor.rs:
```diff
@@ -1194,14 +1219,14 @@ impl Editor {
     }
 
-    /// Closes language servers with timeout. The default timeout is 500 ms, use
+    /// Closes language servers with timeout. The default timeout is 10000 ms, use
```
`10s` seems like a lot and will be noticeable on quit. I think I based 500ms on another implementation

dead10ck replied:
The reason I did this is because the Windows CI was failing because it was timing out shutting down.

More broadly speaking, though, 10s is just the upper limit. If your LS closes in < 500ms, then that's all you'll wait. And we should probably try to give the LS a good chance to shut down gracefully, otherwise we risk potentially corrupting the user's workspace state. #3482 is related.

There are so many LSs out there in use on so many different kinds of hardware that it's inevitable that there will be some that almost always take over 500ms to shut down.

pickfire replied:
I find myself waiting for quite some time for it to save, to the point I lost my patience and just `ctrl-z`.

dead10ck replied:
@pickfire could you elaborate? This is called when shutting down helix. When are you experiencing it taking a long time? Ctrl-Z would just suspend the process and not help anything.

dead10ck replied:
Could you post more details please? Is this happening with this branch and not master? Which LS is this? Do you have any idea why it might be taking 10+ seconds to shut down?

archseer commented on helix-term/src/application.rs:
```diff
@@ -421,6 +406,98 @@ impl Application {
+
+        doc.set_last_saved_revision(doc_save_event.revision);
+
+        let lines = doc.text().len_lines();
+        let bytes = doc.text().len_bytes();
```
This is reporting for the current document version, I'd rather include a clone of the rope on the event and extract the data from there.

dead10ck replied:
Oh that's a good point, will do

archseer commented on helix-term/src/application.rs:
```diff
@@ -421,6 +406,98 @@ impl Application {
+        log::debug!("received editor event: {:?}", event);
+
+        match event {
+            EditorEvent::DocumentSave(event) => {
```
This is after saving is completed, right? Maybe `DocumentSaved` would be a better name?

dead10ck replied:
That's correct. Sure, will do.

archseer commented on helix-term/src/commands.rs:
```diff
@@ -2497,25 +2493,37 @@ enum Modified {
+                    } else if refresh_lang {
+                        let id = doc.id();
+                        doc.detect_language(loader);
+                        let _ = editor.refresh_language_server(id);
```
Does this still run even if there was no format callback? We want to run language detection + refresh if a new file is saved with a name. `:w test.rs` should start highlighting as rust

archseer replied:
Might be better to put this into application.rs, right after `set_path`

dead10ck replied:
No, this code is inside the callback, but that is actually a good point: the write could fail, so doing this here is a bit premature. I agree, I'll move this into the document saved event handler.

dead10ck commented:
I rebased and incorporated a fix with the same effect as #3684. I also did a little refactoring to be able to write a regression test for it; it is now possible to specify a custom TOML string that will get merged into the default languages config to enable testing of language config settings.

I spent a while trying to think of a way to do it in a config format agnostic way, since I know we're not planning on keeping TOML around forever, but I wasn't able to think of any alternative that let us simply override the defaults. Fully specifying the entire language config would be extremely repetitive and get outdated quickly because of the grammar definitions, so this seemed like the worse of the two options.

pickfire commented on helix-term/src/job.rs:
```diff
@@ -93,13 +98,32 @@ impl Jobs {
+                        // clippy doesn't realize this is an error without the derefs
+                        #[allow(clippy::needless_option_as_deref)]
+                        match callback {
+                            Callback::EditorCompositor(call) if compositor.is_some() => {
+                                call(editor, compositor.as_deref_mut().unwrap())
+                            }
+                            Callback::Editor(call) => call(editor),
+
+                            // skip callbacks for which we don't have the necessary references
+                            _ => (),
```
```suggestion
                            Callback::EditorCompositor(call) if compositor.is_some() => {
                                call(editor, compositor.as_deref_mut().unwrap())
                            }
                            Callback::Editor(call) => call(editor),

                            // skip callbacks for which we don't have the necessary references
                            Callback::EditorCompositor(_) => {}
```
Can we make this explicit so we don't miss it the next time we add something to callback?

dead10ck replied:
Good idea 👍

pickfire reviewed (approved):
Still looks good to me but there are conflicts to be resolved.

But looking at the idle timeout, looks like there is a similar pull request exist https://github.com/helix-editor/helix/pull/3172 not sure how well both work together, didn't look closely at the differences.

If anything, probably @archseer would be best to review and merge this.

dead10ck commented:
Thanks a lot @pickfire , I rebased. I don't think that PR is related, though it might touch the same code.

archseer commented on helix-term/src/application.rs:
```diff
@@ -447,6 +427,110 @@ impl Application {
+            let loader = self.editor.syn_loader.clone();
+
+            // borrowing the same doc again to get around the borrow checker
+            let doc = self.editor.document_mut(doc_save_event.doc_id).unwrap();
```
I added a macro for this now: `doc_mut!(self.editor, doc_save_event.doc_id)`

archseer commented on helix-term/src/job.rs:
```diff
@@ -93,13 +98,32 @@ impl Jobs {
+                        #[allow(clippy::needless_option_as_deref)]
+                        match callback {
+                            Callback::EditorCompositor(call) if compositor.is_some() => {
+                                call(editor, compositor.as_deref_mut().unwrap())
```
Why does this need `as_deref_mut()`? What's it being dereferenced into?

dead10ck replied:
I think still just a `&mut Compositor`, but it's needed because just calling `unwrap` moves the original variable, which causes a borrow conflict with the match arm guard.

archseer commented on helix-view/src/document.rs:
```diff
@@ -542,39 +581,137 @@ impl Document {
+    /// and flushing through the queue of pending writes. If any fail,
+    /// it stops early before emptying the rest of the queue.
+    pub async fn close(&mut self) -> Option<DocumentSavedEventResult> {
+        if self.save_sender.is_some() {
+            self.save_sender.take();
         }
```
Should be enough to just call `save_sender.take()` without the `is_some()` check

archseer commented on helix-view/src/document.rs:
```diff
@@ -117,6 +133,9 @@ pub struct Document {
     last_saved_revision: usize,
     version: i32, // should be usize?
     pub(crate) modified_since_accessed: bool,
+    save_sender: Option<UnboundedSender<DocumentSavedEventFuture>>,
+    save_receiver: Option<UnboundedReceiver<DocumentSavedEventFuture>>,
+    current_save: Arc<Mutex<Option<DocumentSavedEventFuture>>>,
```
Note: I still need to spend more time on reviewing how these are used

poliorcetics commented:
I currently encounter crashes infrequently in [`helix-core/src/transaction.rs:397:13`](https://github.com/helix-editor/helix/blob/f4d96b2ca335977ea9b9031e35994b7f15f05c25/helix-core/src/transaction.rs#L397) on the master branch, I wonder if this PR would fix them.

I haven't managed to reproduce it consistently so I cannot test yet

archseer commented on helix-view/src/editor.rs:
```diff
@@ -1269,4 +1305,32 @@ impl Editor {
     }
+
+    pub async fn wait_event(&mut self) -> EditorEvent {
+        let mut saves: FuturesUnordered<_> = self
+            .documents
+            .values_mut()
+            .map(Document::await_save)
+            .collect();
```
This seems expensive to do per event. Could we keep a `saves: FuturesUnordered` on the editor and push an `await_save` onto it when `editor.save` is called?

archseer replied:
That way you also don't need to store `current_save` on the document to prevent it from being dropped

dead10ck replied:
Yeah that's a good idea, will do

archseer replied:
To get this to work we need to add `editor.save`, but I think that's ok since we already do this for `open`/`close`/`switch`

dead10ck replied:
Just thought about this a bit, and I'm afraid it might become problematic. If we just chuck a future into an unordered pool like this every time we do a `:write`, then on quick successive saves of the same file, we could end up doing the saves out of order again. You could end up with this sequence:

1. `:write` 1 adds a save to the document's save queue, and adds the await operation to `editor.saves`
2. `:write` 2 adds a save to the document's save queue, and adds the await operation to `editor.saves`
3. event loop awaits, and happens to choose write 1's future to poll. This pulls write 1 off of the document's internal save queue
4. Part way through writing, the thread is yielded, other stuff happens
5. We go through another iteration of the main event loop. This time the select happens to choose doc 2's write future and polls. This causes the second write to be pulled from the document's internal save queue and begin writing
6. Write 2 finishes
7. Next event loop, write 1 finishes

I think it could basically take us back to square one. Although, maybe a regular `Vec<Future<_>>` could be used? And once the await finishes, pop it off? That way the futures always happen in order, and never get dropped.

archseer replied:
Well there's also a `FuturesOrdered` which acts as a queue: https://docs.rs/futures/latest/futures/stream/struct.FuturesOrdered.html

archseer replied:
Okay I understand this better now. We want a similar setup to the LSP queue: `SelectAll<UnboundedReceiver<_>>`. This way there's no need for `current_save` and `SelectAll` gets to own the `save_receiver` (it'll also get removed from the `SelectAll` if the document ever gets dropped and `save_sender` is closed).

To simplify further I'd also make `async fn save` on the doc return a simple save future and have the `save_sender`/queuing live in a separate `HashMap` on the editor. That way `editor.save` calls `doc.save`, then handles all the queuing and sequential writes.

dead10ck commented on helix-view/src/editor.rs:
```diff
@@ -1095,38 +1114,21 @@ impl Editor {
         };
-
-        // flush out any pending writes first to clear the modified status
-        if let Some(Err(err)) = doc.try_flush_saves().await {
```
Do we not need this any more? I was hitting an issue here before where the document would not close during `write-quit` because the write has to have a chance to be driven to completion and the modified status reset so that the following check didn't fail.

dead10ck replied:
After some testing locally this doesn't appear to be a problem any more. Not sure what was causing it any more. But it's clear now from where it's used that the writes are all flushed before this runs, so it makes to me now.

archseer replied:
Yeah I think that's why I removed this bit 👍🏻

dead10ck commented on helix-view/src/editor.rs:
```diff
@@ -1095,38 +1114,21 @@ impl Editor {
-                log::error!("Error closing doc in language server: {}", err);
-            }
+            // TODO: track error
+            tokio::spawn(language_server.text_document_did_close(doc.identifier()));
```
Are there any consequences you can think of if we don't wait for the did close response?

archseer replied:
This is just a notification we send to the LSP, if there's an error on the server side there's not much we can do other than log the error.

dead10ck commented on helix-view/src/editor.rs:
```diff
@@ -1307,16 +1333,11 @@ impl Editor {
 
-            Some(Some(event)) = saves.next() => {
+            Some(event) = self.save_queue.next() => {
+                self.write_count -= 1;
```
Should we wait until the event is fully handled before considering the save done?

archseer replied:
I was considering it, but if there was an error in the event handling that returned and skipped the `write_count` in the future then `flush_writes` would block. We could make the change though

dead10ck replied:
Yeah I was just realizing that too. I'm actually thinking that no matter what kind of error we hit, we'd probably always want to decrement this count anyway, so let's leave it.

dead10ck commented on helix-view/src/editor.rs:
```diff
@@ -1333,4 +1354,23 @@ impl Editor {
+                        log::error!("error saving document: {}", err);
+                    }
+                };
+                // TODO: if is_err: break?
```
Yeah we should return an error. From an API perspective, I think it makes more sense to let the caller decide how to handle an error. I can take care of this.

dead10ck commented on helix-term/src/application.rs:
```diff
@@ -977,18 +977,7 @@ impl Application {
-                errs.push(err);
-            }
-        }
+        self.editor.flush_writes().await;
```
Note to self: handle this when flush returns an error

dead10ck commented on helix-term/src/commands/typed.rs:
```diff
@@ -79,12 +79,13 @@ fn buffer_close_by_ids_impl(
     doc_ids: &[DocumentId],
     force: bool,
 ) -> anyhow::Result<()> {
+    // TODO: deduplicate with ctx.block_try_flush_writes
```
Note to self: do this

archseer replied:
This is actually a stale TODO, I already did this

archseer replied:
I copied the implementation everywhere but then removed it, forgetting the TODO: https://github.com/helix-editor/helix/pull/2267/commits/aa015cbf156af13625d73516fddddbb36060255a

dead10ck reviewed (commented):
Thanks @archseer ! This is really clever. I left a couple of questions, and notes for myself. I'll take care of the remaining stuff.

dead10ck commented on helix-view/src/editor.rs:
```diff
@@ -1188,6 +1190,30 @@ impl Editor {
+        let path = path.map(|path| path.into());
+        let doc = doc_mut!(self, &doc_id);
+        let future = doc.save(path, force)?;
+        // TODO: if no self.saves for that doc id then bail
```
Note to self: do this

dawkrish commented:
I am not able to write to a file by *helix* but I am able to do with *vim*, what is the reason ?

dawkrish commented:
nevermind, it got automaticaly fixed
